rllib/offline/estimators/fqe_torch_model.py

```diff
@@ -174,37 +182,33 @@ def train_q(self, batch: SampleBatch) -> TensorType:
             self.update_target()
         return losses
 
-    def estimate_q(
-        self,
-        obs: Union[TensorType, List[TensorType]],
-        actions: Union[TensorType, List[TensorType]] = None,
-    ) -> TensorType:
-        """Given `obs`, a list or array or tensor of observations,
-        compute the Q-values for `obs` for all actions in the action space.
-        If `actions` is not None, return the Q-values for the actions provided,
-        else return Q-values for all actions for each observation in `obs`.
-        """
-        obs = torch.tensor(obs, device=self.device)
-        q_values, _ = self.q_model({"obs": obs}, [], None)
-        if actions is not None:
-            actions = torch.tensor(actions, device=self.device, dtype=int)
-            q_values = torch.gather(q_values, -1, actions.unsqueeze(-1)).squeeze(-1)
-        return q_values.detach()
-
-    def estimate_v(
-        self,
-        obs: Union[TensorType, List[TensorType]],
-        action_probs: Union[TensorType, List[TensorType]],
-    ) -> TensorType:
-        """Given `obs`, compute q-values for all actions in the action space
-        for each observations s in `obs`, then multiply this by `action_probs`,
-        the probability distribution over actions for each state s to give the
-        state value V(s) = sum_A pi(a|s)Q(s,a).
-        """
-        q_values = self.estimate_q(obs)
-        action_probs = torch.tensor(action_probs, device=self.device)
+    def estimate_q(self, batch: SampleBatch) -> TensorType:
+        obs = torch.tensor(batch[SampleBatch.OBS], device=self.device)
+        with torch.no_grad():
+            q_values, _ = self.q_model({"obs": obs}, [], None)
+        actions = torch.tensor(
+            batch[SampleBatch.ACTIONS], device=self.device, dtype=int
+        )
+        q_values = torch.gather(q_values, -1, actions.unsqueeze(-1)).squeeze(-1)
+        return q_values
+
+    def estimate_v(self, batch: SampleBatch) -> TensorType:
+        obs = torch.tensor(batch[SampleBatch.OBS], device=self.device)
+        with torch.no_grad():
+            q_values, _ = self.q_model({"obs": obs}, [], None)
+        action_probs = []
+        tmp_batch = batch.copy()
+        for i in range(self.policy.action_space.n):
+            tmp_batch[SampleBatch.ACTIONS] = (
+                np.zeros_like(batch[SampleBatch.ACTIONS]) + i
+            )
+            tmp_probs = torch.exp(
+                compute_log_likelihoods_from_input_dict(self.policy, tmp_batch)
+            )
+            action_probs.append(tmp_probs)
+        action_probs = torch.stack(action_probs).transpose(1, 0)
```

Comment:
**question**: Do these transpose operations work when you have RNNs?

Reply:
Yup, here we're just transposing the batch_size and action_space dimensions.